### PR TITLE
Make T.instance inferable

### PR DESCRIPTION
**Pull request by Keno**

### CODE REVIEW

File: base/compiler/tfuncs.jl [hunk 1]
```diff
@@ -292,8 +293,11 @@ function isdefined_tfunc(@nospecialize(args...))
                 return Const(false)
             elseif !isvatuple(a1) && isbitstype(fieldtype(a1, idx))
                 return Const(true)
-            elseif isa(arg1, Const) && isimmutable((arg1::Const).val)
-                return Const(isdefined((arg1::Const).val, idx))
+            elseif isa(arg1, Const)
+                arg1v = (arg1::Const).val
+                if isimmutable(arg1v) || (isa(arg1v, DataType) && is_dt_const_field(idx))
```
Comment (vtjnash) [inline base/compiler/tfuncs.jl]: could this be:
```julia
if isimmutable(arg1v) || isdefined(arg1v, idx)
```

File: base/compiler/tfuncs.jl [hunk 2]
```diff
@@ -518,13 +522,19 @@ function subtype_tfunc(@nospecialize(a), @nospecialize(b))
 end
 add_tfunc(<:, 2, 2, subtype_tfunc, 0)
 
+is_dt_const_field(fld) = (
```
Comment (vtjnash) [inline base/compiler/tfuncs.jl]: you missed `@nospecialize` here

**Review by vtjnash — COMMENTED** [review]
I don't know—this seems like an implementation detail that we don't want people to depend upon

File: base/compiler/tfuncs.jl (hunk 2 above)
Comment (vtjnash) [reply base/compiler/tfuncs.jl]: Ah, yes. Last time I looked, I think we were missing that check in the caller, but looks like we now ensure it is an Int.

**Comment by vtjnash** [comment]
> Fair, but is that a reason to intentionally make it slower?

Perhaps? It makes some of the code more complex, and as they say "writing code is writing bugs". Also perhaps not as useful to optimize a code smell pattern that we might want to explicitly discourage (since only a subset of objects have this field defined, and the user should be tracking values not Types). But anyways, I won't block the PR over it.

**Review by JeffBezanson — APPROVED** [review]

**Pull request merged**